cpp/mrc/src/internal/service.cpp

```diff
@@ -17,131 +17,295 @@
+
 #include <glog/logging.h>
 
+#include <future>
+#include <mutex>
 #include <ostream>
 #include <utility>
 
 namespace mrc {
 
+Service::Service(std::string service_name) : m_service_name(std::move(service_name)) {}
+
 Service::~Service()
 {
+    if (!m_call_in_destructor_called)
+    {
+        LOG(ERROR) << "Must call Service::call_in_destructor to ensure service is cleaned up before being "
+                      "destroyed";
+    }
+
     auto state = this->state();
     CHECK(state == ServiceState::Initialized || state == ServiceState::Completed);
 }
 
+const std::string& Service::service_name() const
+{
+    return m_service_name;
+}
+
+bool Service::is_service_startable() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state == ServiceState::Initialized);
+}
+
+bool Service::is_running() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state > ServiceState::Initialized && m_state < ServiceState::Completed);
+}
+
+const ServiceState& Service::state() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return m_state;
+}
+
 void Service::service_start()
 {
-    if (forward_state(ServiceState::Running))
+    std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+    if (!this->is_service_startable())
     {
-        do_service_start();
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service has already been started"));
+    }
+
+    if (forward_state(ServiceState::Starting))
+    {
+        // Unlock the mutex before calling start to avoid a deadlock
+        lock.unlock();
+
+        try
+        {
+            this->do_service_start();
+
+            // Use ensure_state here in case the service itself called stop or kill
+            this->ensure_state(ServiceState::Running);
+        } catch (...)
+        {
+            // On error, set this to completed and rethrow the error to allow for cleanup
+            this->forward_state(ServiceState::Completed);
+
+            throw;
+        }
     }
 }
 
 void Service::service_await_live()
 {
-    do_service_await_live();
+    {
+        std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+        if (this->is_service_startable())
+        {
+            throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service must be started before "
+                                                                                     "awaiting live"));
+        }
+
+        // Check if this is our first call to service_await_join
+        if (!m_service_await_live_called)
+        {
+            // Prevent reentry
+            m_service_await_live_called = true;
+
+            // We now create a promise and a future to track the completion of this function
+            Promise<void> live_promise;
+
+            m_live_future = live_promise.get_future();
+
+            // Unlock the mutex before calling await to avoid a deadlock
+            lock.unlock();
```

Comment:
Suggestion: Call `lock.unlock()` early, since no other thread can enter this block.
```suggestion
            // Unlock the mutex before calling await to avoid a deadlock
            lock.unlock();
            
            // We now create a promise and a future to track the completion of this function
            Promise<void> live_promise;

            m_live_future = live_promise.get_future();
```

Reply:
You cant move these around. The lock ensures sequential access to `m_live_future`. Releasing the lock earlier would cause a segfault with 2 concurrent requests to `service_await_live()`.